### PR TITLE
Allow viewing the home page without a cookie or guest user

### PR DESCRIPTION
**Pull request by nb1701**
The index page (which is really a redirect to /programs) used to be behind a login screen. When we removed that, we created a guest user automatically. This is problematic for web crawlers because they can't store cookies, so it will end up redirecting to the callback to create a guest client over and over. Additionally, this approach ends up creating a guest user any time someone loads the site, meaning we get a lot of empty guest accounts.

This changes the approach so that when you load / or /programs, a guest account is not created. It is only created when you click a link to login or start applying to a program.

This also hides the "You're a guest user." text and "End session" link when there is no profile present. On North Star, it hides the banner.

Finally, this moves the DevTools link off to the side so the layout of the other components (the language dropdown) is the same as it is on prod.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

### Instructions for manual testing
Watch the number of accounts in the database, and ensure it isn't incrementing every time you load the home page or do things on that page that don't require an account (language drop down, program category filtering, anything else?).

Try curling the home page directly and ensure it gives you back valid HTML (after the single redirect) and doesn't redirect endlessly (use the `-L` flag).

### Issue(s) this completes

https://github.com/civiform/civiform/issues/8414

### CODE REVIEW

File: server/app/filters/NonUserRoutes.java
Comment (rockycodes) [inline server/app/filters/NonUserRoutes.java:11]: Since we're including `/programs` which is a user route, I think it might be confusing to call this `NonUserRoutes`. I think we should either include `/programs` in the `CiviFormProfileFilter` explicitly or change the name of this. I think other filters also use `NonUserRoutes` (maybe the settings filter?) so I would opt to include `/programs` explicity in `CiviFormProfileFilter` and remove it from this class.

**Comment by rockycodes** [comment]
A thought about "end session" -- could we hide that link if there is no session?

File: server/app/controllers/applicant/ApplicantInformationController.java
Comment (rockycodes) [inline server/app/controllers/applicant/ApplicantInformationController.java:128]: I was thinking that switching the language should actually create a guest user and start a session. We probably want it to switch back to english after 30 minutes of inactivity. What do you think?

File: server/app/views/applicant/NorthStarProgramIndexView.java
Comment (rockycodes) [inline server/app/views/applicant/NorthStarProgramIndexView.java:149]: In North Star, is "review" the program index page? Asking because in our current flow "review" is when you review your answers so we could definitely have an applicant at that point. But I think it's being used differently in North Star.

File: server/app/filters/ValidAccountFilter.java
Comment (rockycodes) [inline server/app/filters/ValidAccountFilter.java:50]: Oh wait this is the other place it's used, not SettingsFilter. I think we wouldn't want `/programs` to be included here since this is only reached if `profile.isPresent` returns true.

File: server/app/views/applicant/ProgramIndexView.java
Comment (rockycodes) [inline server/app/views/applicant/ProgramIndexView.java:133]: This is a nit, but if it's possible to have the comments appear on the same line as the vars that's preferable (maybe use `//` instead of `/*`)

File: server/app/views/applicant/ProgramIndexView.java
Comment (rockycodes) [inline server/app/views/applicant/ProgramIndexView.java:598]: Did you say this would create a new guest account? Why is that?

File: server/app/services/applicant/ApplicantPersonalInfo.java
Comment (rockycodes) [inline server/app/services/applicant/ApplicantPersonalInfo.java:62]: I like it.

File: server/app/views/applicant/ProgramCardViewRenderer.java
Comment (rockycodes) [inline server/app/views/applicant/ProgramCardViewRenderer.java:230]: It's great that those `show` and `review` routes already existed without applicant id!

**Review by rockycodes — COMMENTED** [review]
I think this looks great! As clean an implementation and we could have. Passing them all down as optionals isn't so so bad. I don't have any real red flags except maybe including `/programs` in the `NonUserRoutes`.

File: server/app/filters/NonUserRoutes.java
Comment (rockycodes) [reply server/app/filters/NonUserRoutes.java:11]: I was thinking to include `/` in non user routes since it's not a route a user would see. It just directs them based on their profile if they have one. But I don't feel strongly about it, and the same concern could apply about unintended consequences for other filters using the non user routes filter.

File: server/app/controllers/applicant/ApplicantInformationController.java
Comment (jeffpw-goog) [reply server/app/controllers/applicant/ApplicantInformationController.java:128]: The language is stored in a separate cookie than the session:

image

File: server/app/views/applicant/ApplicantLayout.java
Comment (jeffpw-goog) [inline server/app/views/applicant/ApplicantLayout.java:401]: **suggestion:** I think it should not show the loggedInAsMessage if you don't have a session. I remember when I first used CiviForm I thought it was weird that clicking "End session" had no visible effect, and I think it would make more sense if it didn't show "You're logged in" until you actually start a session.

Same suggestion for the Northstar path; I didn't dig up the code for it.

File: server/app/views/NorthStarBaseView.java
Comment (jeffpw-goog) [reply server/app/views/NorthStarBaseView.java:100]: I am confused - it looks like for a user with no profile this would show "End session" it just wouldn't be linked? Shouldn't it show nothing at all in that case?

I tried to run your PR locally to confirm that's what would happen, but it's showing the link even when I delete the session so I'm confused about that too...

Comment (jeffpw-goog) [reply server/app/views/NorthStarBaseView.java:100]: What would the new message be? I was thinking it wouldn't say anything at all which wouldn't require any translations.

File: server/test/support/ResourceCreator.java
Comment (avaleske) [inline server/test/support/ResourceCreator.java:133]: **non-blocking, nit**: it'd be nice if these names were the same format, so `[insert|new]ActiveTiOnlyProgram`

File: server/test/services/applicant/ApplicantServiceTest.java
Comment (avaleske) [inline server/test/services/applicant/ApplicantServiceTest.java:2101]: **question** do we need to duplicate this test in `ApplicantServiceFastForwardEnabledTest`?

**Review by avaleske — APPROVED** [review]
Thank you for doing this!

Some fast follows that might be nice:
- Fix the positioning of the dev-tools button so it doesn't move the other links out of their prod positions, maybe by absolutely positioning it nearby?
- Do we save the language once the user logs in or starts an application?

File: server/app/views/applicant/ProgramCardsSectionParamsFactory.java
Comment (rockycodes) [inline server/app/views/applicant/ProgramCardsSectionParamsFactory.java:107]: suggestion (non-blocking): You do this same ternary check for `profile.isPresent() && applicantId.isPresent()` in a couple of different places. You could factor it out into a well named method for clarity and ease of reference, but I don't feel strongly about it.

File: server/app/views/applicant/ProgramIndexView.java
Comment (rockycodes) [reply server/app/views/applicant/ProgramIndexView.java:594]: Awesome!

File: server/conf/routes
Comment (rockycodes) [inline server/conf/routes:165]: question (non-blocking): Did anything change about this route? I don't see anything different but github is marking it as changed.

File: server/app/services/applicant/ApplicantService.java
Comment (rockycodes) [reply server/app/services/applicant/ApplicantService.java:1066]: Or even `relevantProgramsWithoutApplicant` which matches names we have for methods elsewhere.

**Review by rockycodes — CHANGES_REQUESTED** [review]
Looks good! Mostly small, non-blocking comments. My biggest blocking comment/question is about the look of the page when `DevTools` isn't rendered.

File: browser-test/image_snapshots/auth_test/guest-just-ended-session.png
Comment (avaleske) [reply browser-test/image_snapshots/auth_test/guest-just-ended-session.png:1]: ahhh I missed that change in the test

**Review by avaleske — APPROVED** [review]

File: browser-test/image_snapshots/applicant_application_index_test/common-intake-form-not-set.png
Comment (rockycodes) [reply browser-test/image_snapshots/applicant_application_index_test/common-intake-form-not-set.png:1]: Ok cool! Looks good. I still would like for Sijia to give the 👍 about this part of the new layout since it will be updating it for all homepages 
image

File: browser-test/image_snapshots/auth_test/user-just-ended-session.png
Comment (rockycodes) [reply browser-test/image_snapshots/auth_test/user-just-ended-session.png:1]: Makes sense. I'll resolve this.

File: server/app/controllers/applicant/ApplicantProgramsController.java
Comment (rockycodes) [reply server/app/controllers/applicant/ApplicantProgramsController.java:183]: Ok cool. If I understand correctly, preventing it from creating a session when filtered means if the previous user did any filtering and then walked away, the next person will see the filtered view. But it's pretty clear in the UI when it's filtered so I don't think that matters. What do you think?

Re: removing the `categories` param from `indexWithoutApplicantId` -- I'm suggesting instead of passing down an empty list everywhere you call `indexWithoutApplicantId`, just pass in an empty list here:
```
ok(
                  programIndexView.renderWithoutApplicant(
                      messagesApi.preferred(request),
                      request,
                      programs,
                      ImmutableList.of()));
```

I don't think this would break anything categories related, but if I'm wrong or if you would rather just leave it as is feel free to resolve this!

**Review by rockycodes — APPROVED** [review]
Lgtm pending approval from Sijia on the nav! Thanks for doing this!

**Pull request merged**
